### PR TITLE
Generalized broadcast_like operator

### PR DESCRIPTION
**Pull request by ifeherva**
## Description ##
This is the generalized implementation of the broadcast_like operator, discussed here:
https://github.com/apache/incubator-mxnet/issues/11871

### CODE REVIEW

File: src/operator/tensor/broadcast_reduce_op.h [hunk 1]
```diff
@@ -137,6 +137,17 @@ struct BroadcastToParam : public dmlc::Parameter<BroadcastToParam> {
   }
 };
 
+struct BroadcastLikeParam : public dmlc::Parameter<BroadcastLikeParam> {
+  TShape input_axes;
+  TShape other_axes;
+  DMLC_DECLARE_PARAMETER(BroadcastLikeParam) {
+    DMLC_DECLARE_FIELD(input_axes).set_default(TShape())
```
[the remark this answers is not in the document]
Comment (ifeherva) [reply src/operator/tensor/broadcast_reduce_op.h]: Well, that's what you proposed :) Can change of course.

**Comment by sbodenstein** [comment]
I find this design counterintuitive:
```
s=mx.sym.broadcast_like(mx.sym.Variable("lhs"), mx.sym.Variable("rhs"), other_axes=(),input_axes=())
s.infer_shape(lhs=(1,7,1,1), rhs=(3,7,9,4))[1]
[(3, 7, 9, 4)]
```
So `input_axes=()` is equivalent to `input_axes=(0, 2, 3)`. Surely an empty tuple should be interpreted as the identity? You could make `other_axes` and `input_axes` optional with default `None`.

**Comment by ifeherva** [comment]
Do you get an error when not passing anything to other_axes and input_axes?

**Comment by ifeherva** [comment]
@sbodenstein There are tests for the empty case.

However, I think passing nothing to the axes params for me means "take the whole" not "take nothing". I find that counterintuitive.

**Comment by sbodenstein** [comment]
Ok, I think we are talking about different cases. Obviously, this PR shouldn't change this case: `mx.sym.broadcast_like()`. It should indeed "take the whole". This case is also being tested for.

But the example I gave was not this, but the case where axes are specified with empty tuples as values: `mx.sym.broadcast_like(lhs_axes=(), rhs_axes=())`. I don't see any tests for this case. And the behaviour seems strange: `lhs_axes=(5,3)` means take two axes, `lhs_axes=(5,)` means take one axis, and `lhs_axes=()` mens take all axes. I'm also happy if it just fails for the empty tuple case.

**Comment by sbodenstein** [comment]
@ifeherva: just checking whether you agree with my comment above.

**Comment by sbodenstein** [comment]
Can't you use `dmlc::optional`?

**Comment by ifeherva** [comment]
Good point, see last commit.

**Comment by sbodenstein** [comment]
Great! Can you just add a test for the empty tuple case?

**Comment by ifeherva** [comment]
Already did:
https://github.com/apache/incubator-mxnet/pull/11984/files#diff-69757562d07268150de8b369ff5b6b61R569

**Comment by sbodenstein** [comment]
Ah, apologies didn't see that. Is it ready to be merged?

**Comment by ifeherva** [comment]
I think so.

**Comment by sbodenstein** [comment]
@szha: could you review this PR?

**Comment by taliesinb** [comment]
@anirudh2290 it would be fantastic if we could get some idea of when a review will be possible.

File: src/operator/tensor/broadcast_reduce_op.h [hunk 2]
```diff
@@ -340,20 +351,60 @@ inline bool BroadcastLikeShape(const nnvm::NodeAttrs& attrs,
...
+    return false;
+  }
+
+  const BroadcastLikeParam& param = nnvm::get<BroadcastLikeParam>(attrs.parsed);
+  TShape oshape;
+
+  // lhs or rhs or both params were not specified
+  if (!param.lhs_axes.has_value() || !param.rhs_axes.has_value()) {
+    CHECK_EQ(lhs_shape.ndim(), rhs_shape.ndim())
     << "Operand of shape " << lhs_shape << " cannot be broadcasted to " << rhs_shape;
```
Comment (szha) [inline src/operator/tensor/broadcast_reduce_op.h]: indentation

File: src/operator/tensor/broadcast_reduce_op.h [hunk 3]
```diff
@@ -340,20 +351,60 @@ inline bool BroadcastLikeShape(const nnvm::NodeAttrs& attrs,
...
+    }
+  } else {
+    auto lhs_axes = param.lhs_axes.value();
+    auto rhs_axes = param.rhs_axes.value();
+
+    CHECK(rhs_axes.ndim() == lhs_axes.ndim())
+    << "Input_axis and other_axis size does not match";
+
+    CHECK(lhs_axes.ndim() > 0)
+    << "Empty axes tuple is not allowed";
```
Comment (szha) [inline src/operator/tensor/broadcast_reduce_op.h]: indentation

File: src/operator/tensor/broadcast_reduce_op.h [hunk 4]
```diff
@@ -340,20 +351,60 @@ inline bool BroadcastLikeShape(const nnvm::NodeAttrs& attrs,
...
+
+    oshape = TShape(lhs_shape);
+    for (index_t i = 0; i < lhs_axes.ndim(); ++i) {
+      auto copyfrom = lhs_axes[i];
+      if (copyfrom < 0) {
+        copyfrom =  lhs_shape.ndim() + copyfrom;
+      }
+      CHECK(copyfrom >= 0 && copyfrom < oshape.ndim())
+      << "Invalid dimension specified in lhs_axes: " << lhs_axes[i];
```
Comment (szha) [inline src/operator/tensor/broadcast_reduce_op.h]: and here

Comment (szha) [reply src/operator/tensor/broadcast_reduce_op.h]: also, either print the whole lhs_axes, or print `i` to make the error message informative.

**Comment by sbodenstein** [comment]
@szha: is it ready to be merged? (I think those indentation and other issues were resolved)

**Pull request merged**